src/LayerEditor/ui/panels/regions_panel.py

```diff
@@ -173,98 +162,104 @@ def __init__(self, layer_heads, parent=None):
-            i_tab += 1
-        while i_tab < self.count():
-            self.removeItem(i_tab)
-            self.tabs.pop(-1)
-            i_tab += 1
-        # Update content.
-        self.setCurrentIndex(self.layer_id_to_idx[self.layer_heads.current_layer_id])
+                tab_widget.reinit(layer_name)
+            # Update content.
+        self.setCurrentIndex(self.layer_heads.current_layer_idx)
 
     def _update_tab_head(self, tab):
         """
         Update tab head, called by tab itself after change of region.
-        :param i:
+        :param tab:
         :return:
         """
         color = tab.region_color
-        item_idx = self.layer_id_to_idx[tab.layer_id]
+        item_idx = self.layer_heads.block.layers.index(tab.layer)
         pixmap = QtGui.QPixmap(16, 16)
         pixmap.fill(color)
         iconPix = QtGui.QIcon(pixmap)
         self.setItemIcon(item_idx, iconPix)
-        self.setItemText(item_idx, "{} ({})".format(tab._layer_name, tab.region.name))
+        self.setItemText(item_idx, "{} ({})".format(tab._layer_name, tab.curr_region.name))
 
     def add_region(self):
         """
         Add new region to all combo and select it in current tab.
         Handle combo changed signal in current tab.
         """
-        dialog = AddRegionDlg(self.layer_heads.max_selected_dim, self.layer_heads.region_names, self)
+        dialog = AddRegionDlg(self.layer_heads.max_selected_dim(), self.layer_heads.region_names, self)
         dialog_result = dialog.exec_()
         if dialog_result == QtWidgets.QDialog.Accepted:
             name = dialog.region_name.text()
             dim = dialog.region_dim.currentData()
-            color = dialog.get_some_color(self.layer_heads.regions._get_available_reg_id() - 1).name()
-            region = self.layer_heads.add_region(name, dim, color)
-            self.current_tab._combo_set_region(region.reg_id)
-            self.update_tabs()
-            self.color_changed.emit()
+            region = self.layer_heads.add_region(name, dim)
+            self.current_tab._combo_set_region(self.current_tab._combo_id_to_idx[region.id])
 
     def remove_region(self):
         """Remove region if it is not assigned to any no shapes"""
-        reg_id = self.layer_heads.selected_region_id
-        self.layer_heads.remove_region(reg_id)
+        reg = self.current_tab.curr_region
+        self.layer_heads.remove_region(reg)
 
     @property
     def current_tab(self):
         """ Current Tab widget. """
-        return self.tabs[self.currentIndex()]
+        return self.widget(self.currentIndex())
 
     def _layer_changed(self):
         """item Changed handler"""
+        assert False, "Not refactored"
         self.layer_heads.select_layer(self.current_tab.layer_id)
-        self.color_changed.emit()
+        self.regions_changed.emit()
+
+    def selection_changed(self):
+        selected = self.layer_heads.block.selection._selected
```

Comment:
I suggest to implement `Selection.shape_ids()` in order to do not operate with graphical items.

Reply:
Implemented `Selection.get_selected_shape_dim_id()` for this purpose. Done.